content/posts/token-driven-processes.md

```diff
@@ -0,0 +1,29 @@
+
+The most sophisticated (ab)use of distributed transaction technology was probably the [WS-AtomicTransaction spec](https://specs.xmlsoap.org/ws/2004/10/wsat/wsat1104.pdf) from 2004, which stood out as too enterprisey even compared to the other WS-* specifications.
+
+The trend seem to have started to change shortly afterwards, around 2007, when Pat Helland published his seminal pager "Life beyond Distributed Transactions: an Apostate's Opinion" where he explained why long-running distributed processes are better realized as a series of message exchanges between stateful entities.
+
+Around the same time messaging frameworks such as NServiceBus and MassTransit became well-known. They both took advantage of the distributed transaction mechanism available in Windows (MS DTC) to ensure exactly-once message processing across a distributed system. The idea was simple. Distributed transactions are bad because they limit autonomy of software components. By limiting the scope of the transaction to just two parties, the messaging node and a database, one can avoid the pitfals while enjoying the benefits. And the benefits were not trivial because having the guarantee that messages don't get duplicated means the business logic code could focus on the actual business task rather than figuring out if a given message has already been processed.
+
+Fast forward 15 years and in today's world distributed system are the norm. Long-running business processes are split into multiple steps. Each step may be executed by different computer under control of a different operating system and implemented in a different programming language. Based on that description one could think that the software industry took Pat Hellands advice seriously and used these frameworks to everyone's benefit. Hardly anything could be farther from truth. The prevalent communication style today is HTTP-based web services and devs continue to struggle with unrelible applications that lose and/or duplicate information.
+
+This could be party explained by the fact that reliable messaging frameworks did not receive support from major software vendors which were preocupied with extending their market shares by selling snake oil in form of deduplication built into the infrastructure e.g. Azure ServiceBus or AWS SQS FIFO. In the meantime technology that supported the reliability of these messaging solutions (like DTC) aged pretty fast. Although a viable replacement in the form of the Outbox pattern has been available for 8 years now, only recently it started to gain attention it deserves (quote Oskar).
```

Comment:
```suggestion
This could be partly explained by the fact that reliable messaging frameworks did not receive support from major software vendors who were preoccupied with extending their market shares by selling snake oil in the form of deduplication built into the infrastructure e.g. Azure ServiceBus or AWS SQS FIFO. Meantime technology that was the bedrock of the reliability for many of the messaging solutions (like DTC) aged pretty fast. Although a viable replacement in the form of the Outbox pattern has been available for 8 years now, only recently has it started to gain the attention it deserves (quote Oskar).
```